### DISCUSSION TITLE
Questions about RNFTs

### DISCUSSION THREAD

**Discussion by Pnin1204** [kodadot/nft-gallery#3675] [Q&A]
Hi, all. I'm Pnin, creating content for the Kusamarian. I'm trying to feature the Reactive NFT standard, and I had some questions that I was hoping could be answered here, much of it clarification.  Would appreciate any information!

---Questions---
Reactive NFTs are programmable NFTs where both on-chain and off-chain data can manipulate the NFT. Is this an accurate statement? Doesn’t that mean off-chain data can possibly be manipulated and cause a reaction in the NFT itself?

The blog states that it is implementation agnostic, but I don't really understand what "chain-independent logic building NFTs" is.

Conditional logic is where, as long as certain conditions are met, the NFTs reacts differently, right? 

Ability to use smart contracts or runtime pallets–just means that it can react to any on-chain logic?

The scheduler is heavily emphasized. Isn’t this accomplishable already with just normal smart contracts and NFTs? Or you mean it’s scheduled as in condition based, as in the example, where if circumstances around an NFT reach certain milestones, then a mint is “scheduled” at those milestones?

What is “a life cycle?” of an NFT? Meaning mint to burn logic?

Generative art already exists. Why the emphasis on RNFTs being able to mint generative art? Is it because the generative logic is on chain?

I don’t understand updating off-chain state. Is it that the NFT can write data to off chain locations?

Writing a novel example was effective, but does that mean that someone could just delete their book and it would happen? Or do people need to accept the change with a signature, similar to accepting a new resource with rmrk?

Royalties were mentioned, and I think that RNFTs will be able to have that royalties programmed into the NFT itself, rather than through a marketplace. Is that correct?

Big question is how this integrates or works with RMRK. I can see the value proposition easily, but dotsama is very focused on RMRK. Is this in conjunction with RMRK? Partnership? Independent? Can RNFTs also be RMRK nfts?

Thanks for any help understanding!

**Comment by vikiival** [comment]
Hey @Pnin1204, 

Thanks for the questions. I'll try to explain them one by one.

**Reactive NFTs are programmable NFTs where both on-chain and off-chain data can manipulate the NFT. Is this an accurate statement?**

Reactive NFTs are the mental model that represents the possibility of changing their state without a to make a transaction initiated by a user.

**Doesn't that mean off-chain data can possibly be manipulated and cause a reaction in the NFT itself?**

The off-chain data can be changed while creating or based on another logic by the Reactive pallet.

**The blog states that it is implementation agnostic, but I don't really understand what "chain-independent logic building NFTs" is.**

Implementation agnostic means that it is possible to use the reactivity for smart-contract-based NFTs and pallet-based NFTs (statemine, basilisk)

**Conditional logic is where, as long as certain conditions are met, the NFTs react differently, right?**

[This is already implemented](https://github.com/kodadot/packages/tree/main/if-that) so you can check. 
The conditional logic should be stored in the JSON NFT metadata and be evaluated by the frontend (e.g. KodaDot). The example could be that the NFT's image should be visible only if the current owner equals your address.

**Ability to use smart contracts or runtime pallets–just means that it can react to any on-chain logic?**

It means the same as "implementation agnostic".

> it can react to any on-chain logic?
Hmm, a kinda cool idea not sure how this would work.


**The scheduler is heavily emphasized. Isn't this accomplishable already with just standard smart contracts and NFTs? Or do you mean it's scheduled as in condition based, as in the example, where if circumstances around an NFT reach certain milestones, then a mint is "scheduled" at those milestones?**


> accomplishable already with just standard smart contracts and NFTs
How? It is impossible to schedule tx in the Ethereum network.

[So far, the only work-in-progress is done by Acala](https://evmdocs.acala.network/network/precompiled-and-predeployed-smart-contracts/schedule)


**What is "a life cycle?" of an NFT? Mean mint to burn logic?**

Throughout your life, you go through many life changes: 
1. You are born
2. Transferred to kindergarten.
3. ...
4. Transferred from university to everyday life
5. ⚰️ 

That means as a person, everyone has a lifecycle, and it's the same with the NFTs. 
They are MINTED, TRANSFERRED, BURNED. These events can hold additional calls executed by pallet or smart contract (based on the implementation).

**Generative art already exists. Why the emphasis on RNFTs being able to mint generative art? Is it because the generative logic is on the chain?**

> Is it because the generative logic is on chain

Yup.

**I don't understand updating the off-chain state. Is it that the NFT can write data to off-chain locations?**

There is a magic called [Substrate off-chain workers](https://www.parity.io/blog/substrate-off-chain-workers-secure-and-efficient-computing-intensive-tasks/). That allows you to do a lot of crazy things such as:

1. fetch the current metadata
2. mutate preferred field
3. pin to IPFS
4. save the new CID to the blockchain

**Writing a novel example was effective, but does that mean that someone could just delete their book and it would happen? Or do people need to accept the change with a signature, similar to accepting a new resource with rmrk?**

> Writing a novel example was effective

Not sure what you mean. Can you provide a link or source? 

**Royalties were mentioned, and I think that RNFTs will be able to have that royalties programmed into the NFT itself, rather than through a marketplace. Is that correct?**

Depends on the final implementation, but both are possible.

**Big question is how this integrates or works with RMRK. I can see the value proposition easily, but dotsama is very focused on RMRK. Is this in conjunction with RMRK? Partnership? Independent? Can RNFTs also be RMRK nfts?**

RMRK is a vertical implementation. 
Vertical means I have NFT. What can I do with that? Add resource, nest it ...


RNFT is horizontal implemetantion.
Horizontal means I have RMRK based NFTs, ERC-721 NFTs, ERC-1155 NFTs and so on. Can I make them reactive?

**Reply by Pnin1204** [reply]
Thanks so much for the reply. It's very helpful. I just had a couple additional questions if that's okay?

So, in the case of a Reactive NFT, since no transaction needs to be initiated by the user, do you foresee any security issues or risks to the user here?

I'm not a dev, so forgive the ignorance, but since smart contracts aren't available on kusama, I would assume reactive NFTs also won't be available on Kusama--only parachains?

In terms of the owner having ultimate control, being able to roll out changes plausibly without user consent such as editing a novel--doesn't that mean that the owner could suddenly decide to destroy their own work entirely without consent of the NFT owner? ie, a disgruntled artist or in a fit of disappointment about their own work, they could "edit" their work and essentially destroy it?

Thanks again!
